lib/stages/VisWriter.cpp

```diff
@@ -0,0 +1,379 @@
+
+        // Check that the number of visibilities matches what we expect
+    } else if (frame.num_prod != acq.num_vis) {
+        FATAL_ERROR("Number of products in frame doesn't match state or file ({:d} != {:d}).",
+                    frame.num_prod, acq.num_vis);
+        return;
+
+    } else {
+
+        // Get the time and frequency of the frame
+        time_ctype t = {std::get<0>(ftime), ts_to_double(std::get<1>(ftime))};
+        uint32_t freq_ind = acq.freq_id_map.at(freq_id);
+
+        // Add all the new information to the file.
+        bool late;
+        double start = current_time();
+
+        // Lock and write data
+        {
+            std::lock_guard<std::mutex> lock(write_mutex);
+            late = acq.file_bundle->add_sample(t, freq_ind, frame);
+        }
+        acq.last_update = current_time();
+        double elapsed = acq.last_update - start;
+
+        DEBUG("Written frequency {:d} in {:.5f} s", freq_id, elapsed);
+
+        // Increase metric count if we dropped a frame at write time
+        if (late) {
+            late_frame_counter.labels({std::to_string(freq_id)}).inc();
+        }
+
+        // Update average write time in prometheus
+        write_time.add_sample(elapsed);
+        write_time_metric.set(write_time.average());
+    }
+}
+
+void VisWriter::init_acq(dset_id_t ds_id, std::map<std::string, std::string> metadata) {
+
+    // Create the new acqState
+    auto fp = datasetManager::instance().fingerprint(ds_id, critical_state_types);
+
+    // If the fingerprint is already known, we don't need to start a new
+    // acquisition, just add a map from the dataset_id to the acquisition we
+    // should use.
+    if (acqs_fingerprint.count(fp) > 0) {
+        INFO("Got new dataset ID={} with known fingerprint={}.", ds_id, fp);
+        acqs[ds_id] = acqs_fingerprint.at(fp);
+        return;
+    }
+
+    INFO("Got new dataset ID={} with new fingerprint={}. Creating new acquisition.", ds_id, fp);
+
+    // If it is not known we need to initialise everything
+    acqs_fingerprint[fp] = std::make_shared<acqState>();
+    acqs[ds_id] = acqs_fingerprint.at(fp);
+    auto& acq = *(acqs.at(ds_id));
+
+    // get dataset states
+    get_dataset_state(ds_id);
+
+    // Check the git version...
+    if (!check_git_version(ds_id)) {
+        acq.bad_dataset = true;
+        return;
+    }
+
+    // TODO: chunk ID is not really supported now. Just set it to zero.
+    uint32_t chunk_id = 0;
+
+    try {
+        acq.file_bundle = std::make_unique<visFileBundle>(
+            file_type, root_path, instrument_name, metadata, chunk_id, file_length, window,
+            kotekan::logLevel(_member_log_level), ds_id, file_length);
+    } catch (std::exception& e) {
+        FATAL_ERROR("Failed creating file bundle for new acquisition: {:s}", e.what());
+    }
+}
+
+VisCalWriter::VisCalWriter(Config& config, const std::string& unique_name,
```

Comment:
@jrs65 I've updated the name of `visCalWriter` to `VisCalWriter`. Where is it actually used?

Reply:
It used to be used where we now use the shared memory buffer. I'm unsure whether we should keep it, or just remove it entirely (@tristpinsm thoughts?). It was a bit of a creative hack to fill a need, and I worry that keeping it around just places restrictions on what we can do with the Writers.

Reply:
If it's no longer used at all I don't see any reason to keep it around. It is only good for that one task

Reply:
I've removed `VisCalWriter` and `visCalFileBundle`.